### PR TITLE
fix Issue 18472: betterC cannot use format at compile time

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
Hmm, -inline causes it to fail.

**Comment by WalterBright** [comment]
Ready to go

File: compiler/test/runnable/test18472.d [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+extern(C) void main()
+{
+    static assert(getData() == "test");
+    static assert("%s %s".format("test", "test") == "test2", "Not working");
+    assert("%s %s".format("test", "test") == "test", "%s %s".format("test", "test"));
+    assert(getData() == "test2", getData());
+}
+
+string getData() {
```
Comment (RazvanN7) [inline compiler/test/runnable/test18472.d:36]: Brace on next line please.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Also, can you target stable please?

**Comment by WalterBright** [comment]
It relies on previous PRs pushed to master.

File: compiler/src/dmd/s2ir.d [hunk 2]
```diff
@@ -164,16 +164,10 @@ private extern (C++) class S2irVisitor : Visitor
 
         if (s.elsebody)
         {
-            const isFalse = s.condition.toBool().hasValue(true);
-            irs.falseBlock += isFalse;
```
Comment (ibuclaw) [inline compiler/src/dmd/s2ir.d:168]: Are you abandoning this folly? So this is now a dead field?  It's unclear how this back-end and the front-end change relate to the issue.

**Comment by ibuclaw** [comment]
@RazvanN7 first 2.103 beta is tomorrow. :sunglasses:

File: compiler/test/runnable/test18472.d (hunk 1 above)
Comment (WalterBright) [reply compiler/test/runnable/test18472.d]: ok

**Comment by dkorpel** [comment]
I know there shouldn't be Phobos dependencies in dmd's test suite, but the code in the linked issue is not fixed.
```D
import std.format;
import std.stdio;
void main()
{
    enum s = "%1$s,%2$s".format("foo","bar");
    writeln(s);
}
```

Still gives

> Error: TypeInfo cannot be used with -betterC

Pointing at `private struct Data` from `std.array.Appender`.

**Comment by maxhaton** [comment]
Given that the testsuite is currently clearly not sufficient for testing the compiler maybe it should explicitly depend on phobos for a while.

**Comment by ibuclaw** [comment]
@maxhaton Phobos has `@betterC` tests.

See https://github.com/dlang/phobos/blob/6d647c582569390ba04b8c172456c8e3134e8c33/std/internal/scopebuffer.d#L132-L136

And https://github.com/dlang/phobos/blob/6d647c582569390ba04b8c172456c8e3134e8c33/posix.mak#L661-L673

**Comment by WalterBright** [comment]
@dkorpel I'll have a look at that one later today. Not surprised there are more cases.

**Comment by WalterBright** [comment]
Remaining problem: https://issues.dlang.org/show_bug.cgi?id=23754

**Pull request merged**